File: apps/app/src/api/skin_manager.rs
use std::{collections::HashMap, path::PathBuf};

use crate::api::Result;
use theseus::{
    prelude::Credentials,
    skin_manager::{Filters, MojangNames, SkinCache, SkinSave},
};

use uuid::Uuid;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("skin-manager")
        .invoke_handler(tauri::generate_handler![
            get_user_skin_data,
            get_heads,
            set_skin,
            delete_skin,
            import_skin,
            check_image,
            check_skin,
            get_cape_data,
            cache_users_skins,
            cache_new_user_skin,
            get_launcher_names,
            save_skin,
            get_skins,
            set_cape,
            get_order,
            save_order,
            get_filters,
            save_filters
        ])
        .build()
}

// Get image size
// invoke('plugin:skin|check_image',path)
#[tauri::command]
pub async fn check_image(path: String) -> Result<bool> {
    Ok(theseus::skin_manager::check_image(path).await?)
}

// Returns true if skin is not in library
// invoke('plugin:skin|check_skin',skin)
#[tauri::command]
pub async fn check_skin(skin: String, id: Uuid) -> Result<bool> {
    Ok(theseus::skin_manager::check_skin(skin, id).await?)
}

// Returns filter options
// invoke('plugin:skin|get_filters')
#[tauri::command]
pub async fn get_filters() -> Result<Filters> {
    Ok(theseus::skin_manager::get_filters().await?)
}

// Saves filter options
// invoke('plugin:skin|save_filters',sort)
#[tauri::command]
pub async fn save_filters(filters: Filters) -> Result<bool> {
    Ok(theseus::skin_manager::save_filters(filters).await?)
}

// Sets player's skin
// invoke('plugin:skin|skin_set_skin', { skin, arms, user })
#[tauri::command]
pub async fn set_skin(
    skin: String,
    arms: String,
    creds: Credentials,
) -> Result<bool> {
    Ok(theseus::skin_manager::set_skin(skin, arms, creds).await?)
}

// Sets the players cape
// invoke('plugin:skin|skin_set_cape', { capeid, token })
#[tauri::command]
pub async fn set_cape(capeid: String, token: String) -> Result<bool> {
    Ok(theseus::skin_manager::set_cape(capeid, token).await?)
}

// Returns cape id or url
// invoke('plugin:skin|skin_set_cape', { cape, key })
#[tauri::command]
pub async fn get_cape_data(cape: String, key: String) -> Result<String> {
    Ok(theseus::skin_manager::get_cape_data(cape, key).await?)
}

// Gets the current account's skin data
// invoke('plugin:skin|skin_get_user_data', { id })
#[tauri::command]
pub async fn get_user_skin_data(id: Uuid) -> Result<SkinCache> {
    Ok(theseus::skin_manager::get_user_skin_data(id).await?)
}

// Makes api request to mojang, updating all accounts' skin data
// invoke('plugin:skin|skin_cache_users_skins')
#[tauri::command]
pub async fn cache_users_skins() -> Result<bool> {
    Ok(theseus::skin_manager::cache_users_skins().await?)
}

// Makes api request to mojang, updating current account's skin data
// invoke('plugin:skin|skin_cache_new_user_skin', { user })
#[tauri::command]
pub async fn cache_new_user_skin(creds: Credentials) -> Result<()> {
    Ok(theseus::skin_manager::cache_new_user_skin(creds).await?)
}

// Saves the skin data to the manager
// invoke('plugin:skin|skin_save_skin', { user, data, name, model, skinid })
#[tauri::command]
pub async fn save_skin(
    user: Uuid,
    data: SkinCache,
    name: String,
    model: String,
    skinid: String,
) -> Result<()> {
    Ok(
        theseus::skin_manager::save_skin(user, data, name, model, skinid)
            .await?,
    )
}

// Updates skin saves from the manager
// invoke('plugin:skin|skin_update_skins', { id })
#[tauri::command]
pub async fn delete_skin(id: Uuid) -> Result<()> {
    Ok(theseus::skin_manager::delete_skin(id).await?)
}

// Gets all account heads from the cache
// invoke('plugin:skin|skin_get_heads')
#[tauri::command]
pub async fn get_heads() -> Result<HashMap<Uuid, String>> {
    Ok(theseus::skin_manager::get_heads().await?)
}

// Gets all saved skins from the manager
// invoke('plugin:skin|skin_get_skins')
#[tauri::command]
pub async fn get_skins() -> Result<Vec<SkinSave>> {
    Ok(theseus::skin_manager::get_skins().await?)
}

// Gets custom skin order
// invoke('plugin:skin|skin_get_order')
#[tauri::command]
pub async fn get_order(user: Uuid) -> Result<Vec<Uuid>> {
    Ok(theseus::skin_manager::get_order(user).await?)
}

// Saves custom skin order
// invoke('plugin:skin|skin_save_order')
#[tauri::command]
pub async fn save_order(order: Vec<Uuid>, user: Uuid) -> Result<()> {
    Ok(theseus::skin_manager::save_order(order, user).await?)
}

// Gets a list of all saved skins in the mojang launcher
// invoke('plugin:skin|skin_get_launcher_names', { path })
#[tauri::command]
pub async fn get_launcher_names(
    path: PathBuf,
    installer: String,
) -> Result<Vec<MojangNames>> {
    Ok(theseus::skin_manager::get_launcher_names(path, installer).await?)
}

// Adds the skin from the mojang launcher to the manager
// invoke('plugin:skin|skin_import_skin', { name, path, user })
#[tauri::command]
pub async fn import_skin(
    id: String,
    path: PathBuf,
    installer: String,
) -> Result<SkinCache> {
    Ok(theseus::skin_manager::import_skin(id, path, installer).await?)
}

File: apps/app-frontend/src/helpers/skin_manager.js
import { invoke } from '@tauri-apps/api/core'
import { SkinViewer } from 'skinview3d'
import { handleError } from '@/store/state.js'
import { ref } from 'vue'

// Returns SkinCache for user from cache
export async function get_user_skin_data(id) {
  return await invoke('plugin:skin-manager|get_user_skin_data', { id })
}

// Sets users skin
export async function set_skin(skin, arms, creds) {
  return await invoke('plugin:skin-manager|set_skin', { skin, arms, creds })
}

// Sets users cape
export async function set_cape(capeid, token) {
  return await invoke('plugin:skin-manager|set_cape', { capeid, token })
}

// Gets skin filter options
export async function get_filters() {
  Filters.value = await invoke('plugin:skin-manager|get_filters')
}

// Saves skin filter options
export async function save_filters() {
  return await invoke('plugin:skin-manager|save_filters', { filters: Filters.value })
}

// Returns true if image is 64x64
export async function check_image(path) {
  return await invoke('plugin:skin-manager|check_image', { path })
}

// Returns true if skin is in library
export async function check_skin(skin, id) {
  return await invoke('plugin:skin-manager|check_skin', { skin, id })
}

// Returns cape info
export async function get_cape_data(cape, key) {
  return await invoke('plugin:skin-manager|get_cape_data', { cape, key })
}

// Stores all users skin data to cache
export async function cache_users_skins() {
  return await invoke('plugin:skin-manager|cache_users_skins')
}

// Caches new user on login
export async function cache_new_user_skin(creds) {
  return await invoke('plugin:skin-manager|cache_new_user_skin', { creds })
}

// Saves SkinCache to the manager
export async function save_skin(user, data, name, model, skinid) {
  return await invoke('plugin:skin-manager|save_skin', { user, data, name, model, skinid })
}

// Deletes the skin save
export async function delete_skin(id) {
  return await invoke('plugin:skin-manager|delete_skin', { id })
}

// Gets custom skin order
export async function get_order(user) {
  return await invoke('plugin:skin-manager|get_order', { user })
}

// Saves custom skin order
export async function save_order(order, user) {
  return await invoke('plugin:skin-manager|save_order', { order, user })
}

// Gets all saved Skins
export async function get_skins() {
  return await invoke('plugin:skin-manager|get_skins')
}

export async function get_heads() {
  account_heads.value = await invoke('plugin:skin-manager|get_heads')
}

export async function get_launcher_names(path, installer) {
  return await invoke('plugin:skin-manager|get_launcher_names', { path, installer })
}

export async function import_skin(id, path, installer) {
  return await invoke('plugin:skin-manager|import_skin', { id, path, installer })
}

export async function get_render(skinData) {
  let arms = skinData.arms
  if (arms == 'classic') arms = 'default'
  const skinViewer = new SkinViewer({
    width: 144,
    height: 144,
    preserveDrawingBuffer: true,
  })
  skinViewer.camera.rotation.x = -0.62
  skinViewer.camera.rotation.y = 0.534
  skinViewer.camera.rotation.z = 0.348
  skinViewer.camera.position.x = -20
  skinViewer.camera.position.y = 10
  skinViewer.camera.position.z = 38

  let skin = skinData.skin
  if (!skin.startsWith('data:image/png;base64,')) skin = tauri.convertFileSrc(skinData.skin)
  await skinViewer.loadSkin(skin, { model: arms })
  const cape = await get_cape_data(skinData.cape, 'url').catch(handleError)
  if (cape !== 'no cape') {
    await skinViewer.loadCape(cape)
  }

  skinViewer.render()
  // delay is required, if not, sometimes the skin isnt loaded before the canvas is saved to img
  await sleep(50)
  const image = skinViewer.canvas.toDataURL()
  skinViewer.dispose()
  return image
}

export const account_heads = ref({})

export const loaded_skins = ref(false)

export const Filters = ref({
  sort: '',
  filter: ''
})

const sleep = (ms = 0) => new Promise((resolve) => setTimeout(resolve, ms))

export const selectedAccount = ref({})
